docs: cds and eds/sds clean up for v2 docs

junr03 opened the pull request.

junr03 commented:
@htuch, I talked to @mattklein123 about the link update. He prefers to leave it out of this PR.

mattklein123 commented on api/cds.proto:
```diff
@@ -174,56 +174,59 @@ message Cluster {
     // The maximum % of an upstream cluster that can be ejected due to outlier
     // detection. Defaults to 10%.
-    google.protobuf.UInt32Value max_ejection_percent = 4;
+    google.protobuf.UInt32Value max_ejection_percent = 4 [(validate.rules).uint32 = {gte:0, lte:100}];
```
nit: gte 0 is redundant as this is uint. same elsewhere.

junr03 replied:
I was debating this one, because I didn't know if we wanted to "catch" that at validate level or the protobuf field level. I have to admit I didn't try to see the failure, but in spirits of eliminating redundancy I see your point. Will update.

mattklein123 commented on api/cds.proto:
```diff
@@ -174,56 +174,59 @@ message Cluster {
     // is detected through consecutive gateway failures. This setting can be
     // used to disable ejection or to ramp it up slowly. Defaults to 0.
-    google.protobuf.UInt32Value enforcing_consecutive_gateway_failure = 11;
+    google.protobuf.UInt32Value enforcing_consecutive_gateway_failure = 11 [(validate.rules).uint32 = {gte:0, lte:100}];
```
nit: flow to 100 col

mattklein123 commented on api/eds.proto:
```diff
@@ -72,8 +74,8 @@ message LbEndpoint {
   // to subset the endpoints considered in cluster load balancing.
-  // TODO(htuch: [V2-API-DIFF] Need to plumb this through Envoy and have
-  // everywhere that canary is used be capable of working on metadata.
+  // [#v2-api-diff: TODO(htuch: [V2-API-DIFF] Need to plumb this through Envoy and have
```
please remove TODOs from docs or put them into a doc comment (comment in comment)

junr03 replied:
I removed as many as possible. For this one I wasn't sure if it should go in a comment tag or a  v2 diff tag. @htuch which one do you want for this?

htuch replied:
Best to make this call on a case-by-case basis. I'd nuke this comment.

junr03 replied:
looks terrible as it is. Will change into doc comment like you suggest.

mattklein123 commented on docs/root/configuration/overview/v2_overview.rst:
```diff
@@ -103,7 +103,7 @@ Mostly static with dynamic EDS
 
 A bootstrap config that continues from the above example with :ref:`dynamic endpoint
-discovery <arch_overview_dynamic_config_sds>` via an EDS gRPC management server listening
+discovery <arch_overview_dynamic_config_sds>` via an :ref:`EDS<envoy_api_file_api/eds.proto>` gRPC management server listening
```
reflow 100  col

mattklein123 commented on docs/root/intro/arch_overview/dynamic_configuration.rst:
```diff
@@ -33,8 +33,8 @@ SDS/EDS only
-mechanism by which Envoy can discover members of an upstream cluster. SDS has been renamed to Endpoint
-Discovery Service (EDS) in the :ref:`v2 API <config_overview_v2>`. Layered on top of a static
+mechanism by which Envoy can discover members of an upstream cluster. SDS has been renamed to :ref:`Endpoint
+Discovery Service (EDS)<envoy_api_file_api/eds.proto>` in the :ref:`v2 API <config_overview_v2>`. Layered on top of a static
```
reflow 100 col

mattklein123 commented:
P.S., thanks, this is awesome.

htuch commented on api/cds.proto:
```diff
@@ -91,26 +90,27 @@ message Cluster {
 
   // The timeout for new network connections to hosts in the cluster.
-  google.protobuf.Duration connect_timeout = 4;
+  google.protobuf.Duration connect_timeout = 4 [(validate.rules).duration.gt.nanos = 0];
```
I would use an empty object here, {}, rather than referencing nanos, since I think this ignores the seconds field in Duration, see https://github.com/google/protobuf/blob/master/src/google/protobuf/duration.proto#L108.

htuch commented on api/cds.proto:
```diff
@@ -148,7 +148,7 @@ message Cluster {
   // cluster types other than STRICT_DNS and LOGICAL_DNS this setting is
   // ignored.
-  google.protobuf.Duration dns_refresh_rate = 16;
+  google.protobuf.Duration dns_refresh_rate = 16 [(validate.rules).duration.gt.nanos = 0];
```
Ditto etc.

junr03 commented:
@mattklein123 updated

htuch commented on docs/root/configuration/cluster_manager/cluster_manager.rst:
```diff
@@ -6,50 +6,14 @@ Cluster manager
-  *(optional, object)* Optional configuration for the cluster discovery service (CDS) API.
+* Cluster manager :ref:`architecture overview <arch_overview_cluster_manager>`
+* :ref:`v1 API reference <config_cluster_manager_v1>`
+* :ref:`v2 API reference <envoy_api_msg_ClusterManager>`
```
FYI, I'm working on `bootstrap.proto` right now, and will take care of the comments on `ClusterManager` in that file.

mattklein123 commented on api/cds.proto:
```diff
@@ -14,6 +14,8 @@ import "google/protobuf/duration.proto";
 import "google/protobuf/struct.proto";
 import "google/protobuf/wrappers.proto";
 
+import "validate/validate.proto";
```
Please make sure all items have some text, even if basic. There is some stuff missing: https://554-90675549-gh.circle-artifacts.com/0/source/generated/docs/api-v2/cds.proto.html. In general I also prefer line breaks between items and comments as it's easier to read for people editing the file.

junr03 commented:
alright, updated. I rendered locally and all the TOC funkiness is addressed. I moved some things around to make the docs more concise, and imo flow better.

htuch commented on api/eds.proto:
```diff
@@ -87,21 +90,24 @@ message LbEndpoint {
   // The limit of 128 is somewhat arbitrary, but is applied due to performance
   // concerns with the current implementation and can be removed when
   // https://github.com/envoyproxy/envoy/issues/1285 is fixed.
```
There are a bunch of messages and URLs in the above two comment blocks. Can you add markup for them?

htuch commented on api/eds.proto:
```diff
@@ -112,7 +118,7 @@ message LocalityLbEndpoints {
   // The limit of 128 is somewhat arbitrary, but is applied due to performance
   // concerns with the current implementation and can be removed when
   // https://github.com/envoyproxy/envoy/issues/1285 is fixed.
```
External link.

htuch commented on api/eds.proto:
```diff
@@ -142,8 +148,10 @@ message LocalityLbEndpoints {
 message EndpointLoadMetricStats {
```
The contents of the above block need some attention. I have a Google internal design proposal for the backend reporting format, I will circulate this for wider design discussion when we have some idea of what makes sense for our needs. Meanwhile, please just remove the example backend load report at the top, and consider providing the aggregated report at the bottom in JSON format.

htuch commented on api/eds.proto:
```diff
@@ -172,8 +180,10 @@ message UpstreamLocalityStats {
+
   // The total number of requests that failed due to errors at the endpoint.
   // For HTTP these are responses with 5xx status codes and for gRPC the
   // grpc-status values {DeadlineExceeded, Unimplemented, Internal,
```
Consider making this a list.

htuch commented on api/eds.proto:
```diff
@@ -205,8 +218,12 @@ message ClusterStats {
+// A load report Envoy sends to the management server.
 message LoadStatsRequest {
-  Node node = 1;  // zone/region where this Envoy runs
+  // zone/region where this Envoy runs
```
"Node identifier for Envoy instance."

htuch commented on api/cds.proto:
```diff
@@ -31,86 +33,108 @@ service ClusterDiscoveryService {
+// specified individually for each defined priority.
 message CircuitBreakers {
+
+  // A Threshold defines CircuitBreaker settings for a RoutingPriority.
```
It's a `Thresholds`

htuch commented on api/cds.proto:
```diff
@@ -31,86 +33,108 @@ service ClusterDiscoveryService {
+
+  //  Thresholds can be defined for each RoutingPriority. If multiple Thresholds
+  // are defined with the same RoutingPriority, the first one in the list is used.
+  // If no Threshold is defined for a given RoutingPriority, the default values
```
Nit: no `Thresholds`. Could also consider linking to the above message def.

htuch commented on api/cds.proto:
```diff
@@ -31,86 +33,108 @@ service ClusterDiscoveryService {
+
+  // If multiple :ref:`Thresholds<envoy_api_msg_CircuitBreakers.Thresholds>` are defined with
+  // the same RoutingPriority, the
+  // first one in the list is used. If no Threshold is defined for a given
```
Nit: Thresholds

htuch commented on api/cds.proto:
```diff
@@ -31,86 +33,108 @@ service ClusterDiscoveryService {
   }
+
+  // If multiple :ref:`Thresholds<envoy_api_msg_CircuitBreakers.Thresholds>` are defined with
+  // the same RoutingPriority, the
```
Suggest adding links to RoutingPriority definition from the two uses in this comment block.

htuch commented on api/cds.proto:
```diff
@@ -31,86 +33,108 @@ service ClusterDiscoveryService {
+
+  // A Thresholds defines CircuitBreaker settings for a RoutingPriority.
   message Thresholds {
+    // The RoutingPriority the specified CircuitBreaker settings apply to.
```
Link to RoutingPriority def.

htuch commented on api/cds.proto:
```diff
@@ -31,86 +33,108 @@ service ClusterDiscoveryService {
   // Supplies the name of the cluster which must be unique across all clusters.
+  // The cluster name is used when emitting
+  // :ref:`statistics <config_cluster_manager_cluster_stats>`.
   // Any : in the cluster name will be converted to _ when emitting statistics.
```
Suggest escaping : and _ here, e.g. `:` and `_`.

htuch commented on api/cds.proto:
```diff
@@ -31,86 +33,108 @@ service ClusterDiscoveryService {
+  // when picking a host in the cluster.
+  LbPolicy lb_policy = 6 [(validate.rules).enum.defined_only = true];
 
   // If the service discovery type is STATIC, STRICT_DNS or LOGICAL_DNS, then
```
Suggest linking to these enum value defs.

htuch commented on api/cds.proto:
```diff
@@ -131,14 +155,18 @@ message Cluster {
     // set so that Envoy will assume that the upstream supports HTTP/2 when
     // making new HTTP connection pool connections. Currently, Envoy only
     // supports prior knowledge for upstream connections. Even if TLS is used
     // with ALPN, http2 must be specified. As an aside this allows HTTP/2
```
Not your change, but this should be reworded to "*http2_protocols_options* must be specified".

htuch commented on api/cds.proto:
```diff
@@ -148,11 +176,9 @@ message Cluster {
   // rate. If this setting is not specified, the value defaults to 5000. For
```
Not your change, but I would s/dns/DNS/ in this block (and add links to the enum values).

htuch commented on api/cds.proto:
```diff
@@ -164,7 +190,10 @@ message Cluster {
-  DnsLookupFamily dns_lookup_family = 17;
+
+  // The DNS IP address resolution policy. If this setting is not specified, the
+  // value defaults to V4_ONLY.
```
Link to enum values in the above blocks. I think everywhere in this file this would be useful.

htuch commented on api/cds.proto:
```diff
@@ -269,7 +315,13 @@ message Cluster {
+    }
+
+    // For each entry, LbEndpoint.Metadata's
     // "envoy.lb" namespace is traversed and a subset is created for each unique
```
Nit: s/"envoy.lb"/*envoy.lb*/

htuch commented on api/cds.proto:
```diff
@@ -306,6 +359,7 @@ message Cluster {
       // to use xxHash now.
```
Link to https://github.com/Cyan4973/xxHash?

htuch commented on api/eds.proto:
```diff
@@ -72,8 +77,6 @@ message LbEndpoint {
   // is "canary", providing the optional canary status of the upstream host.
```
Nit: *canary*, *envoy.lb* formatting rather than "".

htuch commented on api/eds.proto:
```diff
@@ -172,21 +172,32 @@ message UpstreamLocalityStats {
   // originate at the client and the endpoint responded successfuly. For gRPC,
```
Can you fixup the `LoadAssignmentResponse.intervalStats` above to correct field name and add a link to where the field is defined?

htuch commented on api/eds.proto:
```diff
@@ -229,12 +249,17 @@ message ClusterLoadAssignment {
     // [V2-API-DIFF] This is known as maintenance mode in v1.
```
Can you use `[#v2-api-diff: ... ]` here?

htuch commented on api/eds.proto:
```diff
@@ -229,12 +249,17 @@ message ClusterLoadAssignment {
     // [V2-API-DIFF] This is known as maintenance mode in v1.
     double drop_overload = 1;
```
Constraints >=0 && <= 100

htuch commented on api/eds.proto:
```diff
@@ -172,21 +172,32 @@ message UpstreamLocalityStats {
   string cluster_name = 1;
+
   // Need at least one.
   repeated UpstreamLocalityStats upstream_locality_stats = 2;
```
Use constraints here for >= 1 (and change wording so it's not redundant with the constraint).

mattklein123 commented:
@junr03 just ship once @htuch approves. I will look over everything once it's merged and do a follow up PR if I find anything. I would rather move things along. Thanks for all the hard work on this!

junr03 commented:
@mattklein123 sounds good. Sorry it took a couple loops, this one is a pretty hefty move. I rendered locally and took one last pass, I linked and added everything I could think of. I can review a follow up PR if you have one 👌

kyessenov commented on api/cds.proto:
```diff
@@ -31,86 +33,151 @@ service ClusterDiscoveryService {
+// specified individually for each defined priority.
 message CircuitBreakers {
+
+  // A Thresholds defines CircuitBreaker settings for a
```
Remove apostrophe.

kyessenov commented on api/cds.proto:
```diff
@@ -31,86 +33,151 @@ service ClusterDiscoveryService {
     // have the same restrictions as cluster name, i.e. it may be arbritary
     // length.
-    string service_name = 2;
+    string service_name = 2 [(validate.rules).string.min_len = 1];
```
qq: does min_len force it to be required?

htuch replied:
Yep.

junr03 replied:
oops, merged before I got a chance to fix. I talked to @rodaine about this earlier in a different context, and wanted to delete this rule as the field is optional. @htuch can you fix this in your PR?

htuch replied:
Yes, will do.

kyessenov commented on api/eds.proto:
```diff
@@ -219,22 +235,32 @@ message LoadStatsRequest {
 // load_balancing_weight of its Locality.
 message ClusterLoadAssignment {
-  string cluster_name = 1;
+  // Name of the cluster.
```
I think this can be `service_name` from the cluster, too?
